tests/test_downloaderslotssettings.py

```diff
@@ -0,0 +1,75 @@
+import time
+
+from scrapy.crawler import CrawlerRunner
+from scrapy.http import Request
+
+from tests.mockserver import MockServer
+from tests.spiders import MetaSpider
+
+from twisted.internet import defer
+from twisted.trial.unittest import TestCase
+
+
+class DownloaderSlotsSettingsTestSpider(MetaSpider):
+
+    name = 'downloader_slots'
+
+    custom_settings = {
+        "DOWNLOAD_DELAY": 1,
+        "RANDOMIZE_DOWNLOAD_DELAY": False,
+        "DOWNLOAD_SLOTS": {
+            'quotes.toscrape.com': {
+                'concurrency': 1,
+                'delay': 1.5,
+                'randomize_delay': False
+            },
+            'books.toscrape.com': {
+                'delay': 2,
+                'randomize_delay': False
+            }
+        }
+    }
+
+    def start_requests(self):
+        self.times = {None: []}
+
+        slots = list(self.custom_settings.get('DOWNLOAD_SLOTS', {}).keys()) + [None]
+
+        for slot in slots:
+            url = self.mockserver.url(f"/?downloader_slot={slot}")
+            self.times[slot] = []
+            yield Request(url, callback=self.parse, meta={'download_slot': slot})
+
+    def parse(self, response):
+        slot = response.meta.get('download_slot', None)
+        self.times[slot].append(time.time())
+        url = self.mockserver.url(f"/?downloader_slot={slot}&req=2")
+        yield Request(url, callback=self.not_parse, meta={'download_slot': slot})
+
+    def not_parse(self, response):
+        slot = response.meta.get('download_slot', None)
+        self.times[slot].append(time.time())
+
+
+class CrawlTestCase(TestCase):
+
+    def setUp(self):
+        self.mockserver = MockServer()
+        self.mockserver.__enter__()
+        self.runner = CrawlerRunner()
+
+    def tearDown(self):
+        self.mockserver.__exit__(None, None, None)
+
+    @defer.inlineCallbacks
+    def test_delay(self):
+        crawler = CrawlerRunner().create_crawler(DownloaderSlotsSettingsTestSpider)
+        yield crawler.crawl(mockserver=self.mockserver)
+        slots = crawler.engine.downloader.slots
+        times = crawler.spider.times
+        tolerance = 0.3
+
+        delays_real = {k: v[1] - v[0] for k, v in times.items()}
+        error_delta = {k: 1 - delays_real[k] / v.delay for k, v in slots.items()}
```

Comment:
So, for one slot in the tests the delay is 2s, for another it's 1.5s, and it's 1s for the default. Let's say implementation is incorrect, and in all cases delay of 2s is used. error_delta values would be `1 - 2/1.5 = -0.33` and (1-2/1 = 01)`. The assertion below will pass: `max(-0.33, -1) = -0.33 < 0.3` is True.

If the 1.5s delay is used instead of 2s delay, then error_delta values would be `1 - 1.5/2 = 0.25`. `0.25 < 0.3` is True again. It may catch the default 1s download delay though.

Maybe there are reasons these tests may detect an issue (e.g. default DOWNLOAD_DELAY), but I think it could make sense to try improving them. Have you encountered issues with something simpler, e.g. `abs(delays_real[k] - v.delay) < tolerance`?

Reply:
This test implemented directly in the same way as existing test for `DOWNLOAD_DELAY` setting (counting that we need to test multiple downloader slots now)
https://github.com/scrapy/scrapy/blob/80040758236ba1e35a4aa499faf33ed96fc12a92/tests/test_crawl.py#L85-L95

On my local tests `delays_real` calculated from `time.time` calls in spider by some unknown reason always a bit lower that delays from settings:
for delays [1, 1.5, 2] - I received [~0.90, ~1.43, ~ 1.90] in `delays_real`. So on this condition I don't expect issues from value that became negative.
Anyway error calculation method updated to be sure that it will not happen.

I also thought about increasing delay for test from [1, 1.5, 2] to for example [3, 5, 7] in this case with increased delays in test we can safely reduce `tolerance`.

The main question what is acceptable total time for this test?

Reply:
> The main question what is acceptable total time for this test?

The minimum value that still makes the test reliable, both in the sense that it does not break randomly, and that it indeed validates what it is meant to validate. If we need 30 seconds for that, so be it. But if it can be done in 5 seconds, that would be better.